libraries/AP_Scheduler/AP_Scheduler.cpp

```diff
@@ -89,96 +90,87 @@ void AP_Scheduler::tick(void)
  */
 void AP_Scheduler::run(uint32_t time_available)
 {
-    uint32_t run_started_usec = AP_HAL::micros();
-    uint32_t now = run_started_usec;
-
     if (_debug > 1 && _perf_counters == nullptr) {
         _perf_counters = new AP_HAL::Util::perf_counter_t[_num_tasks];
-        if (_perf_counters != nullptr) {
-            for (uint8_t i=0; i<_num_tasks; i++) {
-                _perf_counters[i] = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _tasks[i].name);
-            }
+        for (uint8_t i=0; i < _num_tasks; i++) {
+            _perf_counters[i] = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _tasks[i].name);
         }
     }
     
-    for (uint8_t i=0; i<_num_tasks; i++) {
-        uint16_t dt = _tick_counter - _last_run[i];
-        uint16_t interval_ticks = _loop_rate_hz / _tasks[i].rate_hz;
-        if (interval_ticks < 1) {
-            interval_ticks = 1;
+    uint32_t task_time_buf = AP_HAL::micros();
+    for (uint8_t i = 0; i < _num_tasks; i++) {
+        const uint16_t dt = _tick_counter - _last_run[i];
+        const uint16_t ratio = _loop_rate_hz / _tasks[i].rate_hz;
+        const uint16_t interval_ticks = ratio < 1 ? 1 : ratio;
+        
+        if(dt < interval_ticks) continue;
+        // this task is due to run. Do we have enough time to run it?
+        _task_time_allowed = _tasks[i].max_time_micros;
+
+        // we've slipped a whole run of this task!
+        if (_debug > 4 && dt >= interval_ticks*2) {
+            ::printf("Scheduler slip task[%u-%s] (%u/%u/%u)\n",
+                        (unsigned)i,
+                        _tasks[i].name,
+                        (unsigned)dt,
+                        (unsigned)interval_ticks,
+                        (unsigned)_task_time_allowed);
         }
-        if (dt >= interval_ticks) {
```

Comment:
interval_ticks is unitless and dt is the time. one cannot compare them. I always wondered why this is done and how it can work?